VirtualHavruta/vh.py

```diff
@@ -890,3 +1093,330 @@ def get_topic_descriptions(topic_slugs):
+                candidate_rankings = candidate_rankings[:self.config["database"]["kg"]["max_depth"]]
+            elif len(candidate_chunks) == 0:
+                break
+        retrieval_res_kg = sorted(zip(collected_chunks, ranking_scores_collected_chunks), key=lambda pair: pair[1], reverse=True)
+
+        return retrieval_res_kg,  total_token_count
+
+
+    def get_linker_seed_chunks(self, linker_results: list[dict],
+                        filter_mode: str="primary", msg_id: str="") -> list[Document]:
+        """Given linker results, get the corresponding seed chunks.
+
+        First retrieve the seed nodes: linker results.
+        Find the chunks corresponding to the seed nodes. (1-to-many between nodes and chunks)
+
+        Parameters
+        ----------
+        linker_results
+            results from linker api
+        scripture_query
+            scripture query
+        msg_id
+            message id for slack, optional
+
+        Returns
+        -------
+            list of seed chunks
+        """
+        self.logger.info(f"MsgID={msg_id}. [RETRIEVAL] Starting get_linker_seed_chunks for KG search.")
+        seeds: list[Document] = self.retrieve_nodes_matching_linker_results(linker_results, msg_id, filter_mode=filter_mode)
+        seed_chunks: list[Document] = self.get_chunks_corresponding_to_nodes(seeds)
+        return seed_chunks
+
+
+    def rank_documents(self, chunks: list[Document], enriched_query: str, scripture_query: str|None=None, semantic_similarity_scores: list[float]|None = None,
+                              msg_id: str = "") -> list[Document]:
+        """Rank the document candidates in descending order based on their relevance to the query.
+
+        Return a new list, do not modify the input list.
+
+        Parameters
+        ----------
+        chunks
+            langchain documents
+        enriched_query
+            query enriched with additional context
+        scripture_query
+            query used to retrieve documents from the vector database
+        semantic_similarity_scores
+            optional, provide if available to save some computational costs
+        Returns
+        -------
+            ranked chunks
+        """
+        self.logger.info(f"MsgID={msg_id}. [RETRIEVAL] Starting rank_chunk_candidates for KG search.")
+        total_token_count = 0
+        if not semantic_similarity_scores:
+            if not scripture_query:
+                raise ValueError("Either provide semantic similarity scores or scripture query.")
+            semantic_similarity_scores: np.array = self.compute_semantic_similarity_documents_query(chunks, scripture_query)
+        reference_classes, token_count = self.get_reference_class(chunks, enriched_query)
+        total_token_count += token_count
+        page_rank_scores: np.array = self.get_page_rank_scores(chunks)
+
+        # Combine the scores
+        final_ranking_score = semantic_similarity_scores * reference_classes * page_rank_scores
```

Comment:
Yes, that is an open question, compare: https://github.com/Sefaria/AppliedAI/pull/8#discussion_r1657106836
So I should keep the previous behavior for secondary sources if the context is not KG search, right?

Reply:
Primary references use pagerank scores.
Secondary references don't.

Reply:
For the cases where either all references are primary or secondary, I implemented as commented above.
For the mixed case (in kg search, nodes can be primary or secondary), I still pagerank secondary sources to avoid giving them an advantage in the ranking.